Add prometheus metrics to internal controller

JoelSpeed opened the pull request:
Fixes #119 

Adds prometheus metric server and some metrics to the controller runtime to allow users to see inside the controllers they are building.

Metrics Added:
- Reconcile Errors: Counter of how many reconcile errors have occurred (per controller)
- Queue Length: How many items are in the reconcile queue (per controller)
- Reconcile Time: Histogram metric for how long reconcile's are taking (per controller)

Changes to operation:
- Manager creates a listener in New which binds to the given address (default `:8080`)
- Manager creates a prometheus registry which will be served on the listener
- Manager `.Start()` starts serving the prometheus registry from the manager
- Manager `.Add()` adds metrics from the runnable to the manager's regsitry
- Manager exposes `.AddMetrics()` to add more metrics to the registry
  - This is where kubebuilder consumers will add metrics, in their `.add()` method they can register their metrics with the Manager
- Controller embed `Metrics` struct
  - Metrics struct exposes `GetCollectors()` to allow `Manager.Add()` to register metrics from the controller
  - NewController creates a new set of metrics for use within this particular instance of the controller
  - `processNextWorkItem` updates the metrics within the Controller's Metrics struct

JoelSpeed commented:
/assign @directxman12

JoelSpeed commented:
@DirectXMan12 I'm not following your last comment, sorry. Thanks for getting back quick though!

I've added a more in depth description to the top of this PR which describes the changes I've made, perhaps we could work from that to coordinate our understanding of the problem?

> Is there a good reason to not just have a global or semi-global controller-runtime registry that people can use?

I think the registry within the Manager is semi-global is it not? Any Controller added to the Manager will have it's metrics registered to this instance of the Manager right? Do you mean to make it some exported global within some package so users would import a package and register to that? 

> At the very least, I think it's probably better here to just plumb through a Prometheus registry that things can get a handle to/get injected, and then register metrics against that.

I believe the PR already solves that, the Manager's registry has two points of access, `GetRegistry` and `AddMetrics`, either of which would be reachable within the controller implementations so that kubebuilder users could register metrics to the Manager's registry when they are adding their reconcile function to the controller within their `.add` method. This is how I saw it being used anyhow.

> Is there a compelling use case for the extra abstraction layers on top of that?

Which abstraction are you talking about, I'm lost at this point 😅

DirectXMan12 commented:
The extra abstraction that I'm talking about is the last two bullet points.  I don't see much of a reason to not just have a global controller-runtime prometheus registry (or use the default, maybe?) and if people want to add new metrics, they can just import `sig.k8s.io/controller-runtime/pkg/metrics`, and then `metrics.Registry.MustRegister(...)`.

JoelSpeed commented:
@DirectXMan12 Got ya! I added this extra abstraction after discussing with a colleague about globals vs non-globals. We just went for our preference which is to avoid globals as much as possible but I'm happy to modify the PR to make a global `sigs.k8s.io/controller-runtime/pkg/metrics` if you'd prefer.

The reason I'm suggesting not to use the default prometheus registry is in case you have multiple packages all setting up metrics handling capabilities within one binary. For instance, we plan to run a webhook alongside one of our controllers which sets up it's own metrics endpoint using the global prometheus registry (part of the framework), I'm not personally a fan as at this point, both the webhook and the controllers would serve the same set of metrics if they both used the global registry.

I don't know if that's a common problem, but to me, not using the default prometheus registry seems cleaner, happy to meet in the middle with a controller-runtime registry global if that's what you think is the best approach?

DirectXMan12 commented:
> The reason I'm suggesting not to use the default prometheus registry is in case you have multiple packages all setting up metrics handling capabilities within one binary

Yeah, I agree with that part :-)

> I don't know if that's a common problem, but to me, not using the default prometheus registry seems cleaner, happy to meet in the middle with a controller-runtime registry global if that's what you think is the best approach?

The generally accepted pattern for prometheus is using globals of some variety, AFAIK, so a controller-runtime-global registry might be the best approach.  Then, if people want to serve it off the default endpoint, you should just be able to use `prometheus.Gatherers` to wrap it.

JoelSpeed commented:
@DirectXMan12 I've added a couple of commits;

The first rewrites the metrics registry to be a global as we have previously discussed,

The second adds some tests and modifies the internal metrics to make sure that you can actually register multiple controllers with one manager. The way I was doing it before you'd only ever be able to add one so I've added test cases for this and made sure it all works with multiple controller instances.

DirectXMan12 commented on pkg/manager/internal.go:
```diff
@@ -132,6 +157,29 @@ func (cm *controllerManager) GetRecorder(name string) record.EventRecorder {
+	})
+	mux := http.NewServeMux()
+	mux.Handle("/metrics", handler)
+	server := http.Server{
```
we *might* want to consider using the existing kubernetes machinery for this.  Can be a follow-up PR

DirectXMan12 replied:
yes, please

JoelSpeed commented on pkg/internal/controller/metrics.go:
```diff
@@ -0,0 +1,68 @@
+import "github.com/prometheus/client_golang/prometheus"
+
+// Metrics holds the prometheus metrics used internally by the controller
+type Metrics struct {
```
I was considering the fact that each controller runs in a separate go-routine, perhaps it is better to have one instance of each metric per controller to reduce cross thread communication.

Happy to make the change as you've suggested if you're confident that writing to prometheus metrics is thread-safe? I wasn't sure whether they were or not so went for what I thought was the safer option

DirectXMan12 replied:
yeah, Observe and friends are threadsafe:

> All exported functions and methods are safe to be used concurrently unless specified otherwise.

https://godoc.org/github.com/prometheus/client_golang/prometheus

JoelSpeed commented:
@DirectXMan12 I've changed it all to be globals and also rebased and squashed a bunch of the commits. Does this look more sensible now? 

One other thing I've changed is the metrics servings now has a `0` option to stop the listener being created in the first place, I've used this in tests where it was getting messy. I'm not sure if this is the best way to implement this, any ideas?

droot commented:
@JoelSpeed coming late to the party. Will def. take a look at it tomorrow.

DirectXMan12 commented:
Have a review in progress now that I'm back, but GitHub isn't letting me leave review comments (looks like a temporary blip in their system).  I'll try again this afternoon.

DirectXMan12 commented on pkg/manager/internal.go:
```diff
@@ -157,6 +165,30 @@ func (cm *controllerManager) GetRESTMapper() meta.RESTMapper {
+	// Shutdown the server when stop is closed
+	select {
+	case <-stop:
+		server.Shutdown(context.TODO())
```
`context.Background()`, unless this is actually a TODO

JoelSpeed commented:
@DirectXMan12 I fixed the `context.Background()` thing.

As an FYI, I chose the `0` to disable serving of metrics since this seems to fit with the way other Kubernetes components handle disabling serving

Eg. `--secure-port` on kube-apiserver: 
> The port on which to serve HTTPS with authentication and authorization. If 0, don't serve HTTPS at all.

DirectXMan12 commented:
ack, that makes sense

/approve

fix the rebase issue, then I'll lgtm this

LiliC commented on pkg/internal/controller/metrics/metrics.go:
```diff
@@ -0,0 +1,51 @@
+
+	// ReconcileErrors is a prometheus counter metrics which holds the total
+	// number of errors from the Reconciler
+	ReconcileErrors = prometheus.NewCounterVec(prometheus.CounterOpts{
```
Besides the `ReconcileErrors` I would suggest adding also `ReconsileTotal`. That way we can see if in the past 5mins the rate of errors was too high.

secat replied:
small typo in the suggested new metric name, should probably be `ReconcileTotal`

JoelSpeed replied:
@droot @DirectXMan12 Do you think this would be a worthwhile metric to integrate into this PR?

LiliC commented on Gopkg.lock:
```diff
@@ -581,9 +581,12 @@
+  ]
   pruneopts = "UT"
   revision = "c5b7fccd204277076155f10851dad72b76a49317"
   version = "v0.8.0"
```
There are some nice features in master that are not in this release, maybe now or in the future this could be changed to master instead, as this release is old.

JoelSpeed commented:
@DirectXMan12 I've rebased and squashed the last couple of commits into earlier ones, should be ready to go now

hasbro17 commented:
@JoelSpeed Any update on this? Seems like one of the tests timed out.

JoelSpeed commented:
@hasbro17 Tests are now passing, I had to extend the timeout of the test. I might have missed something with this test but it seems to timeout on the CI but not on my machine, can you see anything obviously wrong with the test?

hasbro17 commented on pkg/manager/manager.go:
```diff
@@ -112,12 +114,17 @@ type Options struct {
 	Namespace string
 
+	// MetricsBindAddress is the TCP address that the controller should bind to
+	// for serving prometheus metrics
```
We should probably mention the default, that metrics will be served on ":8080". 
And that to stop serving metrics this should be set to "0".

I know it's mentioned in `pkg/metrics/listener.go` but this is the main knob that users will tweak, so we should probably clarify that behavior here.

hasbro17 commented:
@JoelSpeed The test seems fine to me. I guess we're waiting for @DirectXMan12 to take another look now.

DirectXMan12 commented on pkg/manager/internal.go:
```diff
@@ -157,6 +165,32 @@ func (cm *controllerManager) GetRESTMapper() meta.RESTMapper {
+func (cm *controllerManager) serveMetrics(stop <-chan struct{}) {
+	handler := promhttp.HandlerFor(metrics.Registry, promhttp.HandlerOpts{
+		ErrorHandling: promhttp.HTTPErrorOnError,
+	})
```
we can follow up with pprof in another PR.  Add an issue for that, though.

droot commented:
@JoelSpeed PR is almost there. Once you add the `dep` stanza for prometheus library and resolve the conflict, we should be good to go.

JoelSpeed commented:
@droot Apologies for the delay. I've resolved the conflict and pinned the prometheus dependency to `v0.9.0` in the `Gopkg.toml`

JoelSpeed commented:
@droot @DirectXMan12 I have been looking at some of our metrics today and realised that, at present, this implementation of the controller-runtime doesn't serve metrics for non-leader pods. Therefore you will see targets down in Prometheus, this is probably worth addressing at some point but wanted to ask whether you'd like me to fix that before merging this or follow up and add a TODO for it

droot commented:
> @droot @DirectXMan12 I have been looking at some of our metrics today and realised that, at present, this implementation of the controller-runtime doesn't serve metrics for non-leader pods. Therefore you will see targets down in Prometheus, this is probably worth addressing at some point but wanted to ask whether you'd like me to fix that before merging this or follow up and add a TODO for it

I am ok with fixing that as a followup PR. Please file an issue for that.

DirectXMan12 commented:
/lgtm
